### ISSUE TITLE
HTTP 429 - include wait time in response

### ISSUE THREAD

**Issue by mmd-osm** [drolbr/Overpass-API#351]
According to [RFC 6585](https://tools.ietf.org/html/rfc6585), the  response representations SHOULD include details explaining the condition, and MAY include a Retry-After header indicating how long to wait before making a new request.

Example:

```
   HTTP/1.1 429 Too Many Requests
   Content-Type: text/html
   Retry-After: 3600

   <html>
      <head>
         <title>Too Many Requests</title>
      </head>
      <body>
         <h1>Too Many Requests</h1>
         <p>I only allow 50 requests per hour to this Web site per
            logged in user.  Try again in xyz seconds</p>
      </body>
   </html>
```

Today, we ask a user to take a look at `/api/status` for further details. This may be fine for tech-savvy people, for normal human beings using applications like overpass turbo it is asking a bit too much  (see respective issue: https://github.com/tyrasd/overpass-turbo/issues/269). Take particular note on the feedback there: 
_Please remember that most users of OT are not experienced coders but just end users who want to click on the button 'Load' & for the program to run. Simple error messages are essential._

Therefore, let's revisit the current /api/status approach!

I'm proposing a few changes:

* Global_Resource_Planner: introduce a new method to return the minimum waiting time, and use this method in dispatcher.cc for command == REQUEST_READ_AND_IDX, right after global_resource_planner.probe.
* In addition to passing the command *Dispatcher::RATE_LIMITED* back to the dispatcher client, we also provide the minimum waiting time.
* Enhance Dispatcher_Client::request_read_and_idx to also retrieve the waiting time in seconds
* Why does a user have to wait 15 seconds, if the quota is (much) larger than 15 seconds. Can't we quit much earlier?
* The rate limit exception should include the minimum waiting time (not only as string, but also as int value).
* Return an HTTP-Header `Retry-After: 3600`, specifying the number of seconds a user has to wait. Also, the current HTML response should include the number of seconds a user has to wait, so this information can be directly shown in overpass turbo, without the additional call to /api/status.

**Comment by drolbr** [comment]
There are two tasks here:
- Refactor the exception classes. I would now base everything on std::exception, and rate limit (as well as bad file content) and so on could get their proper set of parameters.
- Communicate a retry-after value from the dispatcher to the remote client.

Regarding the 15-seconds-rule: This is a convenience feature for browser requests: you can send more requests than rate_limit, and the server will serialize them for you.

**Comment by mmd-osm** [comment]
> Regarding the 15-seconds-rule: This is a convenience feature for browser requests: you can send more requests than rate_limit, and the server will serialize them for you.

Yes, that's clear. Under normal circumstances this makes sense. However, if the rate limit tells you to wait, say 120 seconds, there's not much point blocking the CGI process for 15s. It should rather handle other requests during that time and tell the client to try later more or less immediately.

**Comment by drolbr** [comment]
The problem with the ``Retry-After`` is that in most cases the server does not know either.

Most 429ies are issued because a client starts a request althought another one from that client is still running. The condition is that the other request must finish first (plus quota time), and the server does not know how long that request continues to run.

Thus the options so far are:
* Send always a certain fixed value (e.g. 15 seconds).
* Include the header only if no other request is running and a proper value for retry-after can be implemented. Clients must be prepared for both cases.
* Send a wrong default value in cases where the proper value cannot be determined

**Comment by mmd-osm** [comment]
Some clients are manually parsing /api/status and derive a retry after based on that text file. Providing that very same information in a Rety-After header would be a good improvement already.
That approach has the same issue wrt still running requests, but that’s ok. We should tell the client that it’s futile to try before Retry-After seconds, otherwise we’re not guaranteeing a successful execution. It’s just a minimum mandatory waiting time.
